applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,689 @@
+ */
+
+#include <stdio.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <device.h>
+#include <drivers/uart.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+
+#ifdef CONFIG_BT_LL_NRFXLIB
+#include "ble_controller_hci_vs.h"
+#else
+#error NRFXLIB is required for channel map management
+#endif
+
+#include "chmap_filter.h"
+
+#define MODULE ble_qos
+#include "module_state_event.h"
+#include "ble_event.h"
+#include "config_event.h"
+#include "hid_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BLE_QOS_LOG_LEVEL);
+
+K_SEM_DEFINE(sem_initialized, 0, 1);
+
+struct params_ble {
+	u16_t sample_count_min;
+	u8_t min_channel_count;
+	s16_t weight_crc_ok;
+	s16_t weight_crc_error;
+	u16_t ble_block_threshold;
+	u8_t eval_max_count;
+	u16_t eval_duration;
+	u16_t eval_keepout_duration;
+	u16_t eval_success_threshold;
+} __packed;
+
+struct params_wifi {
+	s16_t wifi_rating_inc;
+	s16_t wifi_present_threshold;
+	s16_t wifi_active_threshold;
+} __packed;
+
+struct params_chmap {
+	u8_t chmap[5];
+} __packed;
+
+struct params_blacklist {
+	u16_t wifi_chn_bitmask;
+} __packed;
+
+static struct chmap_instance *mp_chmap_inst;
+static u8_t m_current_chmap[5] = {0xFF, 0xFF, 0xFF, 0xFF, 0x1F};
+static atomic_t m_processing;
+static atomic_t m_new_blacklist;
+static struct chmap_filter_params *m_new_params;
+static struct k_mutex m_data_access_mutex;
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+# ifndef ROUNDED_DIV
+# define ROUNDED_DIV(a, b)  (((a) + ((b) / 2)) / (b))
+# endif /* ROUNDED_DIV */
+# if DT_NORDIC_NRF_USBD_USBD_0_NUM_IN_ENDPOINTS < 4
+# error Too few USB IN Endpoints enabled.\
+Modify appropriate dts.overlay to increase num-in-endpoints to 4 or more
+# endif
+struct device *cdc_dev;
+static u32_t cdc_dtr;
+
+static void ble_chn_stats_print(bool update_channel_map)
+{
+	char str[64];
+	int str_len, sent_len;
+	s16_t chn_rating;
+	u8_t chn_freq, chn_state;
+	bool buf_loss = false;
+
+# if CONFIG_UART_LINE_CTRL
+	/* Repeated to monitor CDC state */
+	uart_line_ctrl_get(cdc_dev, LINE_CTRL_DTR, &cdc_dtr);
+# endif
+
+	if (!cdc_dtr) {
+		return;
+	}
+
+	if (update_channel_map) {
+		str_len = snprintf(
+			str,
+			sizeof(str),
+			"[%08u]Channel map update\n",
+			(u32_t)k_cycle_get_32());
+		sent_len = uart_fifo_fill(cdc_dev, str, str_len);
+		buf_loss |= sent_len < str_len ? true : false;
+	}
+
+	/* Channel state information print format: */
+	/* BT_INFO={idx0_freq:idx0_state:idx0_rating, ... } */
+
+	str_len = snprintf(str, sizeof(str), "BT_INFO={");
+	sent_len = uart_fifo_fill(cdc_dev, str, str_len);
+	buf_loss |= sent_len < str_len ? true : false;
+
+	for (int i = 0; i < 37; ++i) {
+		chmap_filter_chn_info_get(
+			mp_chmap_inst,
+			i,
+			&chn_state,
+			&chn_rating,
+			&chn_freq);
+		str_len += snprintf(
+						&str[str_len],
+						sizeof(str)-str_len,
+						"%u:%u:%d,",
+						chn_freq,
+						chn_state,
+						chn_rating);
```

Comment:
Is this indentation correct?

Reply:
No, appears to be off by one level. Looked more sensible in vscode. Fixing this.